### PR TITLE
DO-NOT-MERGE [4301] Return provided converter in Protobuf Response Converter Function

### PR DESCRIPTION
**Pull request by resquivel-squareup**
Draft pull request for discussion purposes only

https://github.com/line/armeria/issues/4301

#TODO
------------
Motivation:

Explain why you're making this change and what problem you're trying to solve.

Modifications:

- List the modifications you've made in detail.

Result:

- Closes #<GitHub issue number>. (If this resolves the issue.)
- Describe the consequences that a user will face after this PR is merged.

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/server/annotation/ResponseConverterFunction.java [hunk 1]
```diff
@@ -75,4 +75,23 @@ static <T> T fallthrough() {
...
+        // CompositeResponseConverterFunction, which contains a list of backingConverters
+
+        /* if (defaultResponseConverterFunction.getClass().isInstance(this)) {
+            return this;
+        } else {
+            return defaultResponseConverterFunction;
+        }*/
+
+        return this;
+    }
```
Comment (resquivel-squareup) [inline core/src/main/java/com/linecorp/armeria/server/annotation/ResponseConverterFunction.java:96]: We don't really know at this point (given for example, a `CompositeResponseConverterFunction`) whether the passed-in function is capable of converting.

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/server/annotation/ResponseConverterFunction.java:96]: That's a good question. If the `ResponseConverterFunction` is capable of converting, it returns an `HttpResponse`. If it's not it throw a `FallthroughException` as illustrated here: https://github.com/line/armeria/pull/4312/files#diff-bf99df43890c22eb9efdf47e3ea210cdb632117b76c439c4ceb6f7af49904745R70-R76

So we need to try this and the other:
```java
default ResponseConverterFunction orElse(ResponseConverterFunction other) {
    return new OrElseResponseConverterFunction(this, other);
}

final class OrElseResponseConverterFunction implements ResponseConverterFunction {

    private final ResponseConverterFunction first;
    private final ResponseConverterFunction second;

    OrElseResponseConverterFunction(ResponseConverterFunction first,
                                    ResponseConverterFunction second) {
        this.first = first;
        this.second = second;
    }

    @Override
    public HttpResponse convertResponse(ServiceRequestContext ctx, ResponseHeaders headers,
                                        @Nullable Object result, HttpHeaders trailers) throws Exception {
        try {
            return first.convertResponse(ctx, headers, result, trailers);
        } catch (FallthroughException ignored) {
            return second.convertResponse(ctx, headers, result, trailers);
        }
    }
}


```

File: core/src/main/java/com/linecorp/armeria/server/annotation/ResponseConverterFunction.java [hunk 2]
```diff
@@ -75,4 +76,13 @@ static <T> T fallthrough() {
...
         throw FallthroughException.get();
     }
+
+    /**
+     * Create a ResponseConverterFunction which, during response conversion, will use the current instance
+     * to convert the response. If that fails, the otherConverter will be used
+     * @param otherConverter the ResponseConverterFunction to be used in case the current instance is unable to convert the response
+     */
+    default OrElseResponseConverterFunction orElse(ResponseConverterFunction otherConverter) {
+        return new OrElseResponseConverterFunction(this, otherConverter);
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/annotation/ResponseConverterFunction.java:86]: Let's add
`requireNonNull(otherConverter, "otherConverter");`

File: core/src/test/java/com/linecorp/armeria/internal/server/annotation/OrElseResponseConverterFunctionTest.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
+package com.linecorp.armeria.internal.server.annotation;
```
Comment (minwoox) [inline core/src/test/java/com/linecorp/armeria/internal/server/annotation/OrElseResponseConverterFunctionTest.java:1]: Could you add a copyright?

File: protobuf/src/test/java/com/linecorp/armeria/server/protobuf/ProtobufResponseConverterFunctionProviderTest.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
+package com.linecorp.armeria.server.protobuf;
```
Comment (minwoox) [inline protobuf/src/test/java/com/linecorp/armeria/server/protobuf/ProtobufResponseConverterFunctionProviderTest.java:1]: ditto

File: protobuf/src/main/java/com/linecorp/armeria/server/protobuf/ProtobufResponseConverterFunctionProvider.java [hunk 5]
```diff
@@ -40,7 +40,7 @@ public ResponseConverterFunction createResponseConverterFunction(
             Type returnType,
             ResponseConverterFunction responseConverter) {
         if (isSupportedType(returnType)) {
-            return new ProtobufResponseConverterFunction();
+            return responseConverter.orElse(new ProtobufResponseConverterFunction());
```
Comment (minwoox) [inline protobuf/src/main/java/com/linecorp/armeria/server/protobuf/ProtobufResponseConverterFunctionProvider.java:43]: Could you also do the same thing for `ScalaPbResponseConverterFunctionProvider`?
If you are unfamiliar with Scala, just let us know. We can add a commit for that to this PR. 😉

Comment (minwoox) [reply protobuf/src/main/java/com/linecorp/armeria/server/protobuf/ProtobufResponseConverterFunctionProvider.java:43]: A little optimization we can use is creating a singleton and reusing it. 😄

**Review by minwoox — COMMENTED** [review]
Thanks! This looks good. 😄

File: core/src/test/java/com/linecorp/armeria/internal/server/annotation/OrElseResponseConverterFunctionTest.java (hunk 3 above)
Comment (resquivel-squareup) [reply core/src/test/java/com/linecorp/armeria/internal/server/annotation/OrElseResponseConverterFunctionTest.java:1]: So fast! You reviewed it before I ran the build which reminded me too 😁 I'll add it

File: protobuf/src/test/java/com/linecorp/armeria/server/protobuf/ProtobufResponseConverterFunctionProviderTest.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+        final ResponseConverterFunction
+                function = provider.createResponseConverterFunction(String.class,
+                                                                    new ProtobufResponseConverterFunction());
+
+        assertThat(function).isNull();
+    }
+
+    @Test
+    void requiresOtherProviderToBeNonNull() {
```
Comment (resquivel-squareup) [inline protobuf/src/test/java/com/linecorp/armeria/server/protobuf/ProtobufResponseConverterFunctionProviderTest.java:55]: This passes without any special code

File: core/src/main/java/com/linecorp/armeria/server/annotation/ResponseConverterFunction.java (hunk 2 above)
Comment (resquivel-squareup) [reply core/src/main/java/com/linecorp/armeria/server/annotation/ResponseConverterFunction.java:85]: This passes without any additional code https://github.com/line/armeria/pull/4312/files#r907068352

Looks like it's already configured to be `@NonNullByDefault`

File: protobuf/src/main/java/com/linecorp/armeria/server/protobuf/ProtobufResponseConverterFunctionProvider.java (hunk 5 above)
Comment (resquivel-squareup) [reply protobuf/src/main/java/com/linecorp/armeria/server/protobuf/ProtobufResponseConverterFunctionProvider.java]: > Could you also do the same thing for ScalaPbResponseConverterFunctionProvider?
If you are unfamiliar with Scala, just let us know. 

We're unfamiliar with Scala :)

File: core/src/main/java/com/linecorp/armeria/server/annotation/OrElseResponseConverterFunction.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+ * A {@link ResponseConverterFunction} which takes in two {@link ResponseConverterFunction}s
+ * Upon the call to convert the response, attempt to use the first function.
+ * If the first function is unable to convert the response, fall through to use the second function.
+ */
+final class OrElseResponseConverterFunction implements ResponseConverterFunction {
+
+    private final ResponseConverterFunction firstConverter;
+    private final ResponseConverterFunction secondConverter;
+
+    public OrElseResponseConverterFunction(ResponseConverterFunction firstConverter,
```
[the remark this answers is not in the document]
Comment (resquivel-squareup) [reply core/src/main/java/com/linecorp/armeria/server/annotation/OrElseResponseConverterFunction.java]: Addressed in 946bd041a

File: protobuf/src/test/java/com/linecorp/armeria/server/protobuf/ProtobufResponseConverterFunctionProviderTest.java (hunk 4 above)
Comment (resquivel-squareup) [reply protobuf/src/test/java/com/linecorp/armeria/server/protobuf/ProtobufResponseConverterFunctionProviderTest.java:17]: Addressed in 39c230de

File: core/src/test/java/com/linecorp/armeria/internal/server/annotation/OrElseResponseConverterFunctionTest.java (hunk 3 above)
Comment (resquivel-squareup) [reply core/src/test/java/com/linecorp/armeria/internal/server/annotation/OrElseResponseConverterFunctionTest.java]: Addressed in 39c230de

**Comment by minwoox** [comment]
Superseded by #4336

**Pull request closed without merging**